% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/edgeR-funcs.R
\name{filterByCPM}
\alias{filterByCPM}
\title{Filter EdgeObj and remove lowly expressed genes}
\usage{
filterByCPM(edgeObj, minCPM = 1, minCount = minGroupCount(edgeObj))
}
\arguments{
\item{edgeObj}{An EdgeObject object}

\item{minCPM}{Minimal CPM value, see descriptions below}

\item{mintCount}{Minimal count of samples in which the CPM value is no less than \code{minCPM}

The filter is recommended by the authors of the \code{edgeR} package to remove lowly expressed genes, since including them in differential gene expression analysis will cause extreme differential expression fold-changes of lowly and stochastically expressed genes, and increase false positive rates.

The filter removes genes that are less expressed than 1 copy per million reads (cpm) in at least \code{n} samples, where \code{n} equals the number of samples in the smallest group of the design.}
}
\description{
Filter EdgeObj and remove lowly expressed genes
}
\examples{
myFac <- gl(3,2)
set.seed(1234)
myMat <- matrix(rpois(1200,100), nrow=200, ncol=6)
myMat[1:3,] <- 0
myEdgeObj <- EdgeObject(myMat, 
                       DesignContrast(designMatrix=model.matrix(~myFac),
                        contrastMatrix=matrix(c(0,1,0), ncol=1), groups=myFac),
                        fData=data.frame(GeneSymbol=sprintf("Gene\%d", 1:200)))
myFilteredEdgeObj <- filterByCPM(myEdgeObj)
dim(counts(myEdgeObj))
dim(counts(myFilteredEdgeObj))
## show unfiltered count matrix
dim(counts(myFilteredEdgeObj, filter=FALSE))
}
